lib/rpmug.c

```diff
@@ -1,16 +1,79 @@
 #include "system.h"
 
-#include <pthread.h>
-#include <pwd.h>
-#include <grp.h>
-#include <netdb.h>
+#include <errno.h>
 #include <rpm/rpmlog.h>
 #include <rpm/rpmstring.h>
 
 #include "lib/misc.h"
 #include "lib/rpmug.h"
 #include "debug.h"
 
+#define PWFILE "/etc/passwd"
+#define GRPFILE "/etc/group"
+
+/*
+ * Lookup an arbitrary field based on contents of another in a ':' delimited
+ * file, such as /etc/passwd or /etc/group.
+ */
+static int lookup_field(const char *path, const char *val, int vcol, int rcol,
+			char **ret)
+{
+    int rc = -1; /* assume not found */
+    char *str, buf[BUFSIZ];
+    FILE *f = fopen(path, "r");
+    if (f == NULL) {
+	rpmlog(RPMLOG_ERR, _("failed to open %s for id/name lookup: %s\n"),
+		path, strerror(errno));
+	return rc;
+    }
+
+    while ((str = fgets(buf, sizeof(buf), f)) != NULL) {
+	int nf = vcol > rcol ? vcol : rcol;
+	const char *fields[nf + 1];
+	char *tok, *save = NULL;
+	int col = -1;
+
+	while ((tok = strtok_r(str, ":", &save)) != NULL) {
+	    fields[++col] = tok;
+	    str = NULL;
+	    if (col >= nf)
+		break;
+	}
+
+	if (col >= nf) {
+	    if (rstreq(val, fields[vcol])) {
+		*ret = xstrdup(fields[rcol]);
+		rc = 0;
+	    }
+	}
+    }
+
+    fclose(f);
+
+    return rc;
+}
+
+static int lookup_num(const char *path, const char *val, int vcol, int rcol,
+			long *ret)
+{
+    char *buf = NULL;
+    int rc = lookup_field(path, val, vcol, rcol, &buf);
+    if (rc == 0) {
+	long num = atol(buf);
+	*ret = num;
+	free(buf);
+    }
+    return rc;
+}
+
+static int lookup_name(const char *path, long val, int vcol, int rcol,
+			char **ret)
+{
+    char vbuf[256];
```

Comment:
Where does 256 come from?

Reply:
It's just a "big enough for the purpose" value out of my sleeve, really. There's probably a more appropriate way of expressing "buffer large enough to hold a string representation of a long integer".

Reply:
Is it possible to use [`asprintf`](https://man7.org/linux/man-pages/man3/asprintf.3.html) or is that not portable enough?

Reply:
We have rasprintf() in rpm for portability, we could use that of course. It's just that doing a malloc for converting a static length integer feels so damn ridiculous. Even if it's utterly irrelevant for performance in this case :man_facepalming:

Reply:
I thought about asprintf() but that's not portable.  rasprintf() would work.  Honestly, even though it's silly I say use rasprintf().  Static buffers always lead to future problems, at least in my experience.

Reply:
Yup. It's totally ridiculous how strong that "don't want to waste a malloc on this" feeling is, when in the context it's absolutely nothing at all :smile:

Reply:
Fixed in the last push.